### PR TITLE
Use absolute vfs paths via LLVM `VirtualFileSystem`

### PR DESCRIPTION
**Pull request by jerrymarino**
There are issues when using non absolute-path VFS outside of the
compilation directory, even when setting the `working-directory`. We
can't use absolute paths in this file in Bazel for many reasons. To work
around this, use VFS paths relative to the VFS root via compiler
internals.

Internal to swift and clang - LLVM `VirtualFileSystem` object can
serialize paths relative to the absolute path of the overlay. This
requires the paths are relative to the overlay. While deriving the
in-memory tree roots, it pre-pends the prefix of the `vfsoverlay` path
to each of the entries.

In practice, this is required to use rules_ios vfsoverlay with Xcode's
indexer and SourceKit: e.g. in Bazel Xcode project generators like the
one in rules_ios or XCHammer.

### CODE REVIEW

File: rules/framework/vfs_overlay.bzl [hunk 1]
```diff
@@ -11,11 +11,33 @@ VFSOverlayInfo = provider(
...
+def _get_external_contents(prefix, path_str):
+    return prefix + path_str
+
 # Make roots for a given framework. For now this is done in starlark for speed
 # and incrementality. For imported frameworks, there is additional search paths
 # enabled
 def _make_root(ctx, framework_name, root_dir, extra_search_paths, module_map, hdrs, private_hdrs, has_swift):
     extra_roots = []
+
+    # Compute with the build file path, assuming the build file path is adjacent
```
Comment (jerrymarino) [inline rules/framework/vfs_overlay.bzl:35]: This is a fair assumption. There is one flaw in this plan: the memoization of this vfs will no longer be valid if the entire VFS must be relative

File: rules/framework/vfs_overlay.bzl [hunk 2]
```diff
@@ -161,35 +180,76 @@ def _merge_vfs_infos(base, vfs_infos):
 
 # Internally the "vfs obj" is represented as a dictionary, which is keyed on
 # the name of the root. This is an opaque value to consumers
-def _make_vfs_info(roots):
+def _make_vfs_info(name, data):
     keys = {}
-    for root in roots:
-        name = root["name"]
-        keys[name] = root
+    keys[name] = data
```
Comment (gyfelton) [inline rules/framework/vfs_overlay.bzl:185]: still makes sense to make it a sep. method now that it just do a keyvalue assignment?

File: rules/framework/vfs_overlay.bzl [hunk 3]
```diff
@@ -113,7 +132,7 @@ def _vfs_swift_module_contents(ctx, framework_name, root_dir):
     return {
         "type": "file",
         "name": root_dir + "/" + name,
-        "external-contents": external_contents,
+        "external-contents": _get_external_contents(vfs_prefix, external_contents),
```
Comment (gyfelton) [inline rules/framework/vfs_overlay.bzl:135]: ❤️ having code into simple helper methods for readability

File: rules/framework/vfs_overlay.bzl [hunk 4]
```diff
@@ -161,35 +180,76 @@ def _merge_vfs_infos(base, vfs_infos):
...
+        hdrs = hdrs,
+        private_hdrs = private_hdrs,
+        has_swift = has_swift,
+    )
+
+    vfs_info = _make_vfs_info(framework_name, data)
     if len(merge_vfsoverlays) > 0:
         vfs_info = _merge_vfs_infos(vfs_info, merge_vfsoverlays)
-        roots = vfs_info.values()
+        roots = _roots_from_datas(vfs_parent, vfs_info.values() + [data])
```
Comment (gyfelton) [inline rules/framework/vfs_overlay.bzl:244]: `vfs_info.values() + [data]` is to merge the two arrays correct?

Comment (gyfelton) [reply rules/framework/vfs_overlay.bzl:244]: vfs_info is simply a dict with key being framework_name and value being `data`, now we call `vfs_info.values()` which just returns an array whose only element is `data`, and then we merge it with `[data]?

File: rules/framework/vfs_overlay.bzl (hunk 2 above)
Comment (jerrymarino) [reply rules/framework/vfs_overlay.bzl:185]: The reason we want a factory for the `vfs_info` is that it sets a clear boundary: callers if this method generally aren't concerned about _what_ the details of `vfs_info`. It also improves readability to localizing this logic by related other code. The simple function abstractions here also make it easier to refactor the code

File: rules/framework/vfs_overlay.bzl (hunk 4 above)
Comment (jerrymarino) [reply rules/framework/vfs_overlay.bzl:244]: Yea exactly, you'll also find this in the factory's comment https://github.com/bazel-ios/rules_ios/pull/284/files#diff-450b1330485549712d2fd4e65b2b3fd0a7692fe8f4ec43a0b7b60bbb739313afL163

Comment (gyfelton) [reply rules/framework/vfs_overlay.bzl:244]: Guess my confusion is that it is essentially merging two arrays whose elements are both just `data`? Am I correct? If so is this needed?

**Review by gyfelton — COMMENTED** [review]
Have we tested this against a downstream project? This is potentially breaking change because the overlay_relative is turned on and the path can potentially be very different based on say what the root is. I wonder if we can have a way to compare the generated vfs overlay to 1) always have a clear idea on what the root is 2) to safeguard any regression that changes base of the root and thus breaking things subtly such as debugging ability

File: rules/framework/vfs_overlay.bzl (hunk 4 above)
Comment (jerrymarino) [reply rules/framework/vfs_overlay.bzl:244]: The starlark dictionary function `.values` returns an array, so data is in an array add them together

**Comment by jerrymarino** [comment]
@gyfelton This is a non functional change from the users perspective as far as I have seen. The change happens internally to trigger LLVM creating an absolute path at runtime.  If we compared this Bazel VFS to master, it's quite different but produces the same search. The new one is now compat with Xcode. I can put a followup to dupe the test suite over to VFS.

This PR is for Xcode issues. I'm curious to know how we've been developing and catching regressions on IDE functionality for the rule in general. On master I didn't find tests for indexing and or assertions for IDE functionality. https://github.com/bazel-ios/rules_ios/issues/281 tracks that examples are broken but the tests are green.

**Review by gyfelton — APPROVED** [review]

**Comment by gyfelton** [comment]
@jerrymarino regarding how we test debugger side, it's pretty manual that we have manual test scripts to ensure downstream projects have debugger working. We def. did not pay attention to ensure rules_ios examples have debugger working as no CI job can stop the regression there at this point.

**Comment by jerrymarino** [comment]
@gyfelton makes sense, I suspect getting the tests sorted is larger undertaking and potentially another thread. We could maybe automate LLDB attachment with the python or C++ API. I was thinking of this in other contexts to measure symbolication / attachment times too, which might be a useful overall. At least now we've got the other  issue to noodle on this separately

**Pull request merged**